runtime/lua/vim/lsp/_capability.lua

```diff
@@ -69,9 +77,137 @@ function M:destroy()
   self.active[self.bufnr] = nil
 end
 
+--- Callback invoked when an LSP client attaches.
+--- Use it to initialize per-client state (empty table, new namespaces, etc.),
+--- or issue requests as needed.
+---@param client_id integer
+function M:on_attach(client_id)
+  self.client_state[client_id] = {}
```

Comment:
Client state should generally be stored on the clients (as private `_foo` fields)